### PR TITLE
Read albedo from file over time (v2)

### LINKED ISSUE #243: Time-varying albedo
**Is your feature request related to a problem? Please describe.**
We currently have the capability to use albedo read in from a file. However, the albedo is read only once from the supplied data file, and does not change over time. Here, we want to extend our implementation to read in albedo as it changes over time.

Part of SDI https://github.com/CliMA/ClimaCoupler.jl/issues/276.

Closed by #264, #280

**Describe the solution you'd like**
We can follow the methods that ClimaCoupler.jl uses to read in SST and SIC from files that change over time. This will involve copying over ClimaCoupler's `BCReader` code, which should be moved to a shared location (as should its `Regridder` module, which has also been copied to ClimaLSM.jl.

**Describe alternatives you've considered**
We considered adding this functionality in the coupler instead of ClimaLSM, but ulitmately decided it makes more sense in the land model. This will allow us to use time-varying albedo in uncoupled runs of the land model independent of ClimaCoupler, which we may want to do with, e.g. the canopy model.

**To Do**
- [x] Copy BCReader module from coupler #264 
  - [x] Copy tests
  - [x] Modify source code and tests as needed for ClimaLSM (e.g. don't create CoupledSimulation in tests)
- [x] Add new albedo type in `src/Bucket.jl` - `BulkAlbedoTemporal` #280
  - [x] Use `BCFileInfo` as a backing data structure to store temporal datafile information
  - [x] Rename `BulkAlbedoMap` --> `BulkAlbedoStatic`
- [x] Add a method of `set_initial_parameter_field` for the new albedo type, which calls ~`update_midmonth_data`~ `read_data_fields` and ~`interpolate_midmonth_to_daily`~ `interpolate_data`. #280
  - This is based on the initialization of the sea ice data in the coupler's driver [here](https://github.com/CliMA/ClimaCoupler.jl/blob/main/experiments/AMIP/modular/coupler_driver_modular.jl#L259-L260).
- [x] Add a `update_albedo` function, which will get called from `update_aux` to update `p.bucket.α_sfc` #280
  - [x] For `BulkAlbedoTemporal`, read in the next data point from the file and interpolate if necessary (see coupler drive equivalent [here](https://github.com/CliMA/ClimaCoupler.jl/blob/main/experiments/AMIP/modular/coupler_driver_modular.jl#L416-L430))
  - [x] For other albedo types, return the albedo unchanged
- [x] Add tests for new methods #280

### PR DESCRIPTION
**Pull request by juliasloan25**
## Purpose 
Using the recently-added FileReader module, this PR adds a new albedo model type that allows us to read in albedo from a NetCDF file that varies in time.

Closes #243

## To-do
- [x] Copy infrastructure, tests from #260
- [x] Update `BulkAlbedoTemporal` test to work
  - [x] Add test for `interpolate_data` when current date = a date in the file
    - Maybe add a case to explicitly just return current data

## Notes
- For BulkAlbedoStatic, we use [Regridder.MapInfo](https://github.com/CliMA/ClimaLSM.jl/blob/main/src/shared_utilities/Regridder.jl#L357) to store the albedo dataset file info. For BulkAlbedoTemporal, we use [FileReader.PrescribedData](https://github.com/CliMA/ClimaLSM.jl/blob/main/src/shared_utilities/FileReader.jl#L96). We can probably update BulkAlbedoStatic to use PrescribedData instead, or one of the component structs of PrescribedData (FileInfo?)
  - Maybe in a separate PR we can work on simplify these to just use one struct - see #295 
- This PR introduces a test for `BulkAlbedoTemporal` in `test/standalone/Bucket/albedo_types.jl`, which also tests some functionality of `FileReader.interpolate_data` that isn't tested elsewhere. This test involves a lot of setup (making a PrescribedData object and everything inside of it, reading in a data file for many points in time, creating a space) which is required for the `BulkAlbedoTemporal` test anyway, so it was easier to just write it there. However, we should really add a more thorough test for `interpolate_data` in `test/shared_utilities/file_reader.jl` so the module's functions are thoroughly tested there.
  - Maybe this could be part of a separate PR so it doesn't block this from getting merged - see #285 

----
- [x] I have read and checked the items on the review checklist.

### CODE REVIEW

**Comment by kmdeck** [comment]
you wrote: 
```
For BulkAlbedoStatic, we use [Regridder.MapInfo](https://github.com/CliMA/ClimaLSM.jl/blob/main/src/shared_utilities/Regridder.jl#L357) to store the albedo dataset file info. For BulkAlbedoTemporal, we use [FileReader.PrescribedData](https://github.com/CliMA/ClimaLSM.jl/blob/main/src/shared_utilities/FileReader.jl#L96). We can probably update BulkAlbedoStatic to use PrescribedData instead, or one of the component structs of PrescribedData (FileInfo?)

    Maybe in a separate PR we can work on simplify these to just use one struct
```
this is a great idea.

**Review by kmdeck — APPROVED** [review]

File: src/standalone/Bucket/Bucket.jl [hunk 1]
```diff
@@ -75,40 +79,111 @@ struct BulkAlbedoFunction{FT} <: AbstractLandAlbedoModel{FT}
...
+        t_start::FT,
+        surface;
+        comms = ClimaComms.SingletonCommsContext(),
+        varname = "sw_alb"
+    ) where {FT}
+
+Constructor for the BulkAlbedoStatic that implements a default albedo map, `comms` context, and value for `α_snow`.
+The `varname` must correspond to the name of the variable in the NetCDF file specified by `bcfile_path`.
+
+The `cesm2_albedo_dataset_path` artifact can be used as a default with this type.
```
[the remark this answers is not in the document]
Comment (juliasloan25) [reply src/standalone/Bucket/Bucket.jl:153]: good idea!

File: src/standalone/Bucket/Bucket.jl [hunk 2]
```diff
@@ -455,17 +563,63 @@ function make_update_aux(model::BucketModel{FT}) where {FT}
...
+"""
+    update_albedo!(model_albedo::BulkAlbedoTemporal, parameters, Y, p, t)
+
+Update the surface albedo for time t. For a file containing albedo
+information over time, this reads in the value for time t.
+"""
+function next_albedo(model_albedo::BulkAlbedoTemporal, parameters, Y, p, t)
+    # Get the current date from `t`
+    sim_info = model_albedo.albedo_info.sim_info
+    date = to_datetime(sim_info.date_ref + Second(sim_info.t_start) + Second(t))
```
[the remark this answers is not in the document]
Comment (juliasloan25) [reply src/standalone/Bucket/Bucket.jl:614]: good catch! It errors if t isn't rounded to the nearest whole number. I'll change this to round the times.

File: src/standalone/Bucket/Bucket.jl [hunk 3]
```diff
@@ -455,17 +563,63 @@ function make_update_aux(model::BucketModel{FT}) where {FT}
...
+    # Use next date if it's closest to current time
+    if date >= to_datetime(next_date_in_file(model_albedo.albedo_info))
```
[the remark this answers is not in the document]
Comment (juliasloan25) [reply src/standalone/Bucket/Bucket.jl:616]: yes exactly. should I add a comment to make this more clear?

File: test/standalone/Bucket/albedo_types.jl [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+                atmosphere = bucket_atmos,
+                radiation = bucket_rad,
+            )
+            # Initial conditions with no moisture
+            Y, p, coords = initialize(model)
+            Y.bucket.T .= init_temp.(coords.subsurface.z, FT(280.0))
+            Y.bucket.W .= 0.0
+            Y.bucket.Ws .= 0.0
+            Y.bucket.σS .= 0.0
+            set_initial_aux_state! = make_set_initial_aux_state(model)
```
[the remark this answers is not in the document]
Comment (juliasloan25) [reply test/standalone/Bucket/albedo_types.jl:108]: yeah, good idea

File: src/standalone/Bucket/Bucket.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (kmdeck) [reply src/standalone/Bucket/Bucket.jl:616]: maybe? I think it's ok now, but might be useful for someone new looking at it.

**Pull request merged**